### ISSUE TITLE
Struggling with serial port string communication

### ISSUE THREAD

**Issue by tbeu** [modelica-3rdparty/Modelica_DeviceDrivers#373]
@tbeu 
Dear Thomas
Yes it worked! We finally could test it and it worked perfectly, thank you so much : )

Now I am struggling with sending and receiving strings.
When I send a command e.g. "DL31" the board should e.g. turn the led on and then send back "DL31".

I am not able to receive strings in dymola from the board.
For my test example I use an **Arduino Uno** board (for now using the standard 8bit data) with this sketch:

```
void setup() {
  // Serial.begin(38400, SERIAL_7E1);
  Serial.begin(38400, SERIAL_8E1);
}

void loop() {

  Serial.write("Hallo\0");
  delay(1000);
}
```

and the following **modelica model**:

![ModelicaModel_ReceiveString](https://user-images.githubusercontent.com/32899658/231580339-7c448f6d-3725-4751-973c-894b295ccba1.PNG)

![ModelicaModel_ReceiveString_PortSettings](https://user-images.githubusercontent.com/32899658/231583478-7174dec6-c829-465a-9c86-fc3c9904e3eb.PNG)

**Modelica code**:
```
model Test_ReceiveString_06
  extends Modelica.Icons.Example;

  Modelica_DeviceDrivers.Blocks.Communication.SerialPortReceive serialPortReceive(
    enableExternalTrigger=true,
    autoBufferSize=false,
    userBufferSize=6,
    Serial_Port="COM4",
    baud=Modelica_DeviceDrivers.Utilities.Types.SerialBaudRate.B38400,
    parity=1) "to send and receive data "
    annotation (Placement(transformation(extent={{-50,50},{-30,30}})));
  Modelica.Blocks.Logical.Not not1 annotation (Placement(transformation(
        extent={{-10,-10},{10,10}},
        rotation=180,
        origin={20,60})));
  Modelica.Blocks.Sources.BooleanPulse booleanPulse(period(displayUnit=
          "s") = 2)
               annotation (Placement(transformation(
        extent={{-10,-10},{10,10}},
        rotation=180,
        origin={60,60})));
  Modelica_DeviceDrivers.Utilities.TriggeredPrint triggeredPrint(message=
        getString.data)
    annotation (Placement(transformation(extent={{-50,72},{-30,92}})));
  GetString                                                        getString(
      dataBitSize=8, bufferSize=6)
    annotation (Placement(transformation(extent={{-10,-30},{10,-10}})));
equation 
  connect(triggeredPrint.trigger, not1.y) annotation (Line(points={{-40,76},
          {-40,60},{9,60}},                 color={255,0,255}));
  connect(booleanPulse.y, not1.u)
    annotation (Line(points={{49,60},{32,60}}, color={255,0,255}));
  connect(not1.y, serialPortReceive.trigger) annotation (Line(points={{9,60},
          {-40,60},{-40,52}}, color={255,0,255}));
  connect(serialPortReceive.pkgOut, getString.pkgIn)
    annotation (Line(points={{-29.2,40},{0,40},{0,-9.2}}, color={0,0,0}));
  annotation (Icon(coordinateSystem(preserveAspectRatio=false, extent={{-140,
            -120},{140,120}})),Diagram(coordinateSystem(preserveAspectRatio=
            false, extent={{-140,-120},{140,120}})),
    experiment(StopTime=30, __Dymola_Algorithm="Dassl"));
end Test_ReceiveString_06;

```

I get the following **error**:

![ModelicaError_ReceiveString](https://user-images.githubusercontent.com/32899658/231580938-df8c11f7-8cc8-4a3d-91f3-d6448d984634.PNG)

I studied all the examples within the MDD Library but still not understanding how to make it work.
Must all strings be null terminated with `\0` or could they also be terminated with carriage return `\r` or new line feed `\n` and so on?
Could you help me with this too?
Should I create a new issue for this?

Vielen Dank und beste Grüsse!

_Originally posted by @sertho in https://github.com/modelica-3rdparty/Modelica_DeviceDrivers/issues/368#issuecomment-1505978002_

**Comment by tbeu** [comment]
When checking some questions ([SO](https://stackoverflow.com/q/10989364/8520615) or [Arduino.cc](https://forum.arduino.cc/t/serial-write-null-character/135210)), can you please try

```
Serial.write("Hallo\0", 6);
```
or
```
Serial.write("Hallo");
Serial.write((byte)0);
```

**Comment by tbeu** [comment]
> Must all strings be null terminated with `\0` or could they also be terminated with carriage return `\r` or new line feed `\n` and so on?

Yes, `MDD_SerialPackagerGetString` requires a null-terminated buffer, see

https://github.com/modelica-3rdparty/Modelica_DeviceDrivers/blob/8ac7971cf17aa4d8102dff30312266ff51ef428e/Modelica_DeviceDrivers/Resources/Include/MDDSerialPackager.h#L457-L495

**Comment by sertho** [comment]
> When checking some questions ([SO](https://stackoverflow.com/q/10989364/8520615) or [Arduino.cc](https://forum.arduino.cc/t/serial-write-null-character/135210)), can you please try
> 
> ```
> Serial.write("Hallo\0", 6);
> ```
> 
> or
> 
> ```
> Serial.write("Hallo");
> Serial.write((byte)0);
> ```

Both suggestions seem to work : )
Even if I activate `autoBufferSize` within the `serialPortReceive` component.
I will go on on this and let you know if it works to build-up a serial communication using only strings.
Thank you!

**Comment by sertho** [comment]
> > Must all strings be null terminated with `\0` or could they also be terminated with carriage return `\r` or new line feed `\n` and so on?
> 
> Yes, `MDD_SerialPackagerGetString` requires a null-terminated buffer, see
> 
> https://github.com/modelica-3rdparty/Modelica_DeviceDrivers/blob/8ac7971cf17aa4d8102dff30312266ff51ef428e/Modelica_DeviceDrivers/Resources/Include/MDDSerialPackager.h#L457-L495


What if my machine's controler returns not null terminated but CR or LF terminated strings?

Could I then adopt the header file [Modelica_DeviceDrivers/Modelica_DeviceDrivers/Resources/Include/MDDSerialPackager.h](https://github.com/modelica-3rdparty/Modelica_DeviceDrivers/blob/8ac7971cf17aa4d8102dff30312266ff51ef428e/Modelica_DeviceDrivers/Resources/Include/MDDSerialPackager.h#L457-L495) ?

Is this also a header-only library?

**Comment by tbeu** [comment]
@sertho Can this be closed by now?

**Comment by sertho** [comment]
Please don't close it yet. I did'nt solve it by now.
Could you please help me with this:

> > > Must all strings be null terminated with `\0` or could they also be terminated with carriage return `\r` or new line feed `\n` and so on?
> > 
> > 
> > Yes, `MDD_SerialPackagerGetString` requires a null-terminated buffer, see
> > https://github.com/modelica-3rdparty/Modelica_DeviceDrivers/blob/8ac7971cf17aa4d8102dff30312266ff51ef428e/Modelica_DeviceDrivers/Resources/Include/MDDSerialPackager.h#L457-L495
> 
> What if my machine's controler returns not null terminated but CR or LF terminated strings?
> 
> Could I then adopt the header file [Modelica_DeviceDrivers/Modelica_DeviceDrivers/Resources/Include/MDDSerialPackager.h](https://github.com/modelica-3rdparty/Modelica_DeviceDrivers/blob/8ac7971cf17aa4d8102dff30312266ff51ef428e/Modelica_DeviceDrivers/Resources/Include/MDDSerialPackager.h#L457-L495) ?
> 
> Is this also a header-only library?


Thank you!

**Comment by tbeu** [comment]
@bernhard-thiele Actually, there is an inconsistency between `MDD_SerialPackagerAddString` and its counterpart `MDD_SerialPackagerGetString`: Whereas `MDD_SerialPackagerAddString` supports sending non-null-terminated strings (due to the used `memcpy` instead of `strncpy`), `MDD_SerialPackagerGetString` always expects null-terminated strings. What do you think of always adding a trailing null in `MDD_SerialPackagerAddString` in the sending buffer and in `MDD_SerialPackagerGetString` for the receiving buffer, if not found? I could create PR if you are fine with the proposed changes.

**Comment by bernhard-thiele** [comment]
@tbeu Sounds good. Null-termination is granted for Modelica Strings, but this would make the API more robust if it is used from custom C code.

**Comment by tbeu** [comment]
> What if my machine's controler returns not null terminated but CR or LF terminated strings?

PR #377 improves the output buffer handling by always adding a trailing null character instead of raising an error and terminating the simulation. You are then able to manually manipulate that string e.g. by means of the [Modelica.Utilities.Strings](https://doc.modelica.org/Modelica%204.0.0/Resources/helpDymola/Modelica_Utilities_Strings.html#Modelica.Utilities.Strings) functions.

**Comment by sertho** [comment]
Hi

I could finally test your new implementation. But it does not work.
I am not sure if the error is on my side or is due to the implementation #377 .

For testing I downloaded the masterbranch yesterday.

I'm still not able to receive strings in dymola from the board.
For my test example I use an Arduino Uno board with the following sketch
(now using the new  byte size parameter for the 7bit data introduced with #378 ) :

```

void setup() {
  Serial.begin(38400, SERIAL_7E1); // using 7bit
  // Serial.begin(38400, SERIAL_8E1);
}

void loop() {
  
  // -------------------------------------------------------------------------------------------------------
  
  // -> this does not work
  // Dymola Output:
  // Warning: MDDSerialPackager.h: MDD_SerialPackagerGetString warning.
  // No terminating '\0' found in buffer of bufferSize: 4
  Serial.write("abc");
  
  // -------------------------------------------------------------------------------------------------------
  
  // -> this works, if in Modelica the parameter "getString.bufferSize" = 4 ("a", "b", "c" + "\0")
  // Serial.write("abc\0", 4);

  // -------------------------------------------------------------------------------------------------------
  
  // -> this works, if in Modelica the parameter "getString.bufferSize" = 4 ("a", "b", "c" + "(byte)0")
  // Serial.write("abc");
  // Serial.write((byte)0);
  
  // -------------------------------------------------------------------------------------------------------
  
  delay(1000);
}
```

and the following modelica model:

![grafik](https://github.com/modelica-3rdparty/Modelica_DeviceDrivers/assets/32899658/62be1ef6-ed2f-416c-8bff-421a73862903)

Modelica code:
```
model Test_ReceiveString_07
  extends Modelica.Icons.Example;

  Modelica_DeviceDrivers.Blocks.Communication.SerialPortReceive serialPortReceive(
    enableExternalTrigger=true,
    autoBufferSize=true,
    Serial_Port="COM6",
    baud=Modelica_DeviceDrivers.Utilities.Types.SerialBaudRate.B38400,
    parity=1,
    byteSize=7)
              "to send and receive data "
    annotation (Placement(transformation(extent={{-50,30},{-30,10}})));
  Modelica.Blocks.Sources.BooleanPulse booleanPulse(period(displayUnit="s")
       = 1, startTime=1)
               annotation (Placement(transformation(
        extent={{-10,-10},{10,10}},
        rotation=180,
        origin={60,40})));
  Modelica_DeviceDrivers.Utilities.TriggeredPrint triggeredPrint(message=
        getString.data)
    annotation (Placement(transformation(extent={{30,-10},{50,-30}})));
  Modelica_DeviceDrivers.Blocks.Packaging.SerialPackager.GetString
    getString(bufferSize=4)
    annotation (Placement(transformation(extent={{-10,-30},{10,-10}})));
equation 
  connect(booleanPulse.y, serialPortReceive.trigger) annotation (Line(
        points={{49,40},{-40,40},{-40,32}}, color={255,0,255}));
  connect(serialPortReceive.trigger, triggeredPrint.trigger)
    annotation (Line(points={{-40,32},{-40,40},{40,40},{40,-14}},
                                                 color={255,0,255}));
  connect(getString.pkgIn, serialPortReceive.pkgOut)
    annotation (Line(points={{0,-9.2},{0,20},{-29.2,20}}, color={0,0,0}));
  annotation (Icon(coordinateSystem(preserveAspectRatio=false, extent={{-140,
            -120},{140,120}})),Diagram(coordinateSystem(preserveAspectRatio=
            false, extent={{-140,-120},{140,120}}), graphics={Text(
          extent={{-80,-20},{0,-60}},
          textColor={28,108,200},
          horizontalAlignment=TextAlignment.Left,
          textString="bufferSize must match,
otherwise it does not work.")}),
    experiment(StopTime=5, __Dymola_Algorithm="Dassl"));
end Test_ReceiveString_07;

```


If I understand it right, this https://github.com/modelica-3rdparty/Modelica_DeviceDrivers/issues/373#issuecomment-1667328587 means that it is not necessary anymore to add a null terminator from the board, so this should work, right?
```
Serial.write("abc");
```

I get this Dymola output:
![grafik](https://github.com/modelica-3rdparty/Modelica_DeviceDrivers/assets/32899658/ad7f0d70-74b8-4b28-8c17-32024be2bca6)


If instead of `Serial.write("abc");` for the Arduino code I use this:
```
  Serial.write("abc");
  Serial.write((byte)0);
```
then the Dymola output is what I would like to get:
![grafik](https://github.com/modelica-3rdparty/Modelica_DeviceDrivers/assets/32899658/748389c0-480a-4aba-9927-13e6637a0446)

(I don't know why, but the first event at 1s is always missed, but not a problem for me...)

Just as recapitulation, why I am working on this:
My machine's controler returns not null terminated but CR or LF terminated strings `\r\n`?
I was able to send commands and receive answers using Python serial library and now I want to realize this directly from Dymola using MDD.
An example for my use case is this:
I send "T6" to the machine and a get back a string e.g. "T6001500014F". I can then convert this string to meaningful outputs. In this case "T" = Temperature, "6" = number of sensor, "0" = sensor status, "0150" = Hexvalue of Temperature, etc...
This is why I need to be able to receive not null-terminated strings and also strings of different sizes.

For our purpuse, we developed a bidirectional serial port component in Modelica using MDD.
It works as long as we use integers... but not yet for our case of using strings for the communication.
![grafik](https://github.com/modelica-3rdparty/Modelica_DeviceDrivers/assets/32899658/af2375a8-60ac-4bdf-9ac2-beb23a9eba74)


In case you are interested to incorporate this block into MDD, here is the code (it may need to be adjusted, enhanced...):
```
block SerialPortBidirectional_7bit
  "A block for sending and receiving serial datagrams using the serial interface"
  extends Modelica_DeviceDrivers.Utilities.Icons.SerialPortIcon;
  //   extends Modelica_DeviceDrivers.Blocks.Communication.Internal.PartialSampleTrigger;
  import Modelica_DeviceDrivers.Packaging.SerialPackager;
  import Modelica_DeviceDrivers.Packaging.alignAtByteBoundary;
  import Modelica_DeviceDrivers.Communication.SerialPort;
  import Modelica_DeviceDrivers.Utilities.Types.SerialBaudRate;
  parameter Boolean autoBufferSize = true
    "true, buffer size is deduced automatically, otherwise set it manually"
    annotation(Dialog(group="Incoming data"), choices(checkBox=true));
  parameter Integer userBufferSize=16*64
    "Buffer size of message data in bytes (if not deduced automatically)" annotation(Dialog(enable=not autoBufferSize, group="Incoming data"));
  parameter String Serial_Port="/dev/ttyPS1" "Serial port to send data"
   annotation (Dialog(group="Incoming data"));
  parameter SerialBaudRate baud= SerialBaudRate.B9600 "Serial port baud rate"
  annotation (Dialog(group="Incoming data"));
  parameter Integer parity = 0 "set parity (0 - no parity, 1 - even, 2 - odd)"
    annotation (Dialog(group="Outgoing data"));
  parameter Integer byteSize(min=5, max=8) = 8 "Number of data bits transmitted per (serial data format) byte (8 is most common)"
    annotation (Dialog(group="Outgoing data"));
  Modelica_DeviceDrivers.Blocks.Interfaces.PackageOut pkgOut(
    pkg=SerialPackager(if autoBufferSize then bufferSize else userBufferSize),
    dummy(start=0, fixed=true))
    annotation (Placement(transformation(
        extent={{-20,-20},{20,20}},
        rotation=90,
        origin={108,0})));
  Modelica_DeviceDrivers.Blocks.Interfaces.PackageIn pkgIn
    annotation (
      Placement(transformation(
        extent={{-20,-20},{20,20}},
        rotation=270,
        origin={-108,0})));
  Modelica.Blocks.Interfaces.BooleanInput triggerReceive
    annotation (Placement(transformation(
        extent={{-20,-20},{20,20}},
        rotation=90,
        origin={-40,-120})));
  Modelica.Blocks.Interfaces.BooleanInput triggerSend
    annotation (Placement(transformation(
        extent={{-20,-20},{20,20}},
        rotation=90,
        origin={40,-120})));

protected 
  Integer bufferSize;
  SerialPort sPort = SerialPort(Serial_Port, if autoBufferSize then bufferSize else userBufferSize, parity, receiver, baud, byteSize);
  parameter Integer receiver = 1 "Set to be a receiver port";
  Real dummy(start=0, fixed=true);

equation 

  // Output port -> Receiver
  when initial() then
    bufferSize = if autoBufferSize then alignAtByteBoundary(pkgOut.autoPkgBitSize) else userBufferSize;
  end when;
  pkgOut.trigger = triggerReceive "using inherited trigger";
  when pkgOut.trigger then
    pkgOut.dummy = Modelica_DeviceDrivers.Blocks.Communication.Internal.DummyFunctions.readSerial(
      sPort,
      pkgOut.pkg,
      time);
  end when;

  // Input port -> Send
  when initial() then
    pkgIn.userPkgBitSize = if autoBufferSize then -1 else userBufferSize*8;
    pkgIn.autoPkgBitSize = 0;
    //bufferSize = if autoBufferSize then Modelica_DeviceDrivers.Packaging.SerialPackager_.getBufferSize(pkgIn.pkg) else userBufferSize;
  end when;
  pkgIn.backwardTrigger = triggerSend "using inherited trigger";
  when pkgIn.trigger then
    dummy = Modelica_DeviceDrivers.Blocks.Communication.Internal.DummyFunctions.sendToSerial(
      sPort,
      pkgIn.pkg,
      bufferSize,
      pkgIn.dummy);
  end when;

  annotation (preferredView="info",
          Icon(coordinateSystem(preserveAspectRatio=false, extent={{-100,
            -100},{100,100}}), graphics={Text(extent={{-150,136},{150,96}},
            textString="%name"),         Text(extent={{-150,82},{150,42}},
          textString="%Serial_Port"),            Text(extent={{-152,-48},{148,-88}},
          textString="%baud")}), Documentation(info="<html>
<h4>Support for receiving datagrams over a serial port</h4>
<h4>Example</h4>
<p>
See <a href=\"modelica://Modelica_DeviceDrivers.Blocks.Examples.TestSerialPackager_SerialPort\"><code>TestSerialPackager_SerialPort</code></a>.
</p>
</html>"));
end SerialPortBidirectional_7bit;

```

Thank you for helping!
Let me know, if I should do other tests for debugging!

Beste Grüsse,
Thomas

**Comment by tbeu** [comment]
> If I understand it right, this [#373 (comment)](https://github.com/modelica-3rdparty/Modelica_DeviceDrivers/issues/373#issuecomment-1667328587) means that it is not necessary anymore to add a null terminator from the board, so this should work, right?

Argh, an empty string buffer was returned for non null terminated string data. This is now addressed by #382.
